### ISSUE TITLE
[NEWS] GPSR - umożliwiamy dodanie informacji o bezpieczeństwie produktu / GPSR - we allow you to add product safety information

### ISSUE THREAD

**Issue by MaciejFrackowiak** [allegro/allegro-api#10210]
**13 grudnia 2024** zacznie obowiązywać nowe rozporządzenie Unii Europejskiej [**General Product Safety Regulation**](https://commission.europa.eu/business-economy-euro/doing-business-eu/eu-product-safety-and-labelling/product-safety/general-product-safety-regulation_en). Jednym z jego elementów są załączniki, w których sprzedający powinien umieścić informacje z ostrzeżeniami o produkcie np. instrukcja produktu. Jeśli sprzedający nie ma instrukcji, może to być zdjęcie oznaczeń na opakowaniu produktu. Ważne, aby załączniki były w języku urzędowym kraju, z którego jest kupujący. Zatem, jeśli sprzedający udostępnia swoją ofertę na wszystkie rynki Allegro, instrukcja powinna zawierać wersję polską, czeską, słowacką, czy węgierską.

**Jakie zmiany wprowadziliśmy?**

Dla:

* [POST /sale/product-offers](https://developer.allegro.pl/documentation/#operation/createProductOffers),
* [PATCH /sale/product-offers/{offerId}](https://developer.allegro.pl/documentation/#operation/editProductOffers)

dodaliśmy nowy obiekt “**productSet.[].safetyInformation**”, w którym wskażesz, czy dla danego produktu posiadasz informacje o bezpieczeństwie produktu. Dla każdego produktu musisz przekazać **osobne** dane.

Obecnie możesz zdefiniować jeden z dwóch typów informacji:

* "**ATTACHMENTS**" - jeżeli dodajesz załączniki jako informacje o bezpieczeństwie produktu. Do każdego produktu możesz dodać maksymalnie 20 załączników.
* “**NO_SAFETY_INFORMATION**” - jeśli deklarujesz, że produkt nie posiada informacji o bezpieczeństwie. 

**Uwaga!** W przyszłości umożliwimy dodanie informacji o bezpieczeństwie produktu również w postaci opisu tekstowego (type: “TEXT”).

**Przykładowy request:**
```json5
{
  "productSet": [
    {
      "safetyInformation": {
         "type": "ATTACHMENTS",                               // typ informacji, obecnie dostępne wartości: ATTACHMENTS, NO_SAFETY_INFORMATION
         "attachments": [
               {
                 "id": "184e6fcc-db46-4727-b80c-2bb652d9b58b" // identyfikator załącznika (“SAFETY_INFORMATION_MANUAL”)
               }
         ]
      },
      ...
    }
  ],
  ...
}
```
> 

Pole “**safetyInformation**” aktualnie jest opcjonalne. W dniu **13.12.2024** oznaczymy je jako **obowiązkowe**.

Jednocześnie, dla załączników z informacją o bezpieczeństwie produktu (“**SAFETY_INFORMATION_MANUAL**”), które dodasz w oparciu o zasób [POST /sale/offer-attachments](https://developer.allegro.pl/documentation/#operation/createOfferAttachmentUsingPOST) wprowadziliśmy dziś zmiany:

* oprócz dotychczasowych plików w formacie PDF możesz dodać pliki graficzne w formatach JPEG, JPG, PNG;
* rozmiar pojedynczego pliku może wynosić maksymalnie 50 MB.

**Dlaczego wprowadziliśmy zmianę?**

Wdrożyliśmy informacje o bezpieczeństwie produktu, aby dostosować Allegro do prawa Unii Europejskiej. GPSR, oprócz producentów, importerów, dystrybutorów, itd., nakłada nowe obowiązki na platformy internetowe umożliwiające sprzedaż na odległość. Platformy muszą dostosować i zorganizować swój interfejs w taki sposób, aby sprzedający mogli wypełnić swoje obowiązki wynikające z rozporządzenia.

-----

On **December 13, 2024**, the new European Union Regulation [**General Product Safety Regulation**](https://commission.europa.eu/business-economy-euro/doing-business-eu/eu-product-safety-and-labelling/product-safety/general-product-safety-regulation_en) will come into effect. One of its elements are attachments, in which the seller should include information with warnings about the product, such as product instructions. If the seller does not have a manual, it can be a photo of the product labelings. The attachments must be in the official language of the buyer's country. Therefore, if a seller offers their product on all Allegro markets, the manual should include Polish, Czech, Slovak, and Hungarian versions.

**What changes have we made?**

For:

* [POST /sale/product-offers](https://developer.allegro.pl/documentation/#operation/createProductOffers),
* [PATCH /sale/product-offers/{offerId}](https://developer.allegro.pl/documentation/#operation/editProductOffers)

we have added a new object “**productSet.[].safetyInformation**”, in which you will indicate whether you have product safety information for a given product. You must pass **separate** data for each product.

Currently, you can define one of two types of information:

* "**ATTACHMENTS**" - if you are adding attachments as product safety information. You can add a maximum of 20 attachments to each product.
* “**NO_SAFETY_INFORMATION**” - if you declare that the product has no safety information.

**Note!** In the future, we will make it possible to add product safety information also in the form of a text description (type: “TEXT”).

**Sample request:**
```json5
{
  "productSet": [
    {
      "safetyInformation": {
         "type": "ATTACHMENTS",                               // type of information, currently available values: ATTACHMENTS, NO_SAFETY_INFORMATION
         "attachments": [
               {
                 "id": "184e6fcc-db46-4727-b80c-2bb652d9b58b" // attachment identifier (“SAFETY_INFORMATION_MANUAL”)
               }
         ]
      },
      ...
    }
  ],
  ...
}
```
> 

The “**safetyInformation**” field is currently optional. On **December 13, 2024**, we will mark it as **required**.

At the same time, for attachments with product safety information (“**SAFETY_INFORMATION_MANUAL**”) that you add based on the [POST /sale/offer-attachments](https://developer.allegro.pl/documentation/#operation/createOfferAttachmentUsingPOST) we have made changes today:

* in addition to the existing PDF files, you can add image files in JPEG, JPG, PNG formats;
* the size of a single file can be up to 50 MB.

**Why have we made the change?**

We have implemented product safety information to align Allegro with European Union law. The GPSR, in addition to producers, importers, distributors, etc., imposes new obligations on online platforms that enable distance selling. Platforms must adjust and organize their interface in such a way that sellers can fulfill their obligations under the regulation.

**Comment by tomasz-grzesik** [comment]
A co z załącznikami typu Instrukcja z informacją o bezpieczeństwie w polu attachments?
Nie wystarczy tam wrzucić taki załącznik?
![obraz](https://github.com/user-attachments/assets/47da7d11-7f8d-4626-a490-dee9174cdf48)

**Comment by MaciejFrackowiak** [comment]
Zgodnie z regulacją informacja powinna być przypisana do każdego produktu wchodzącego w skład oferty. Za pośrednictwem pola "attachments" możesz dodać tylko po jednym typie załącznika do samej oferty, nie produktu.

**Comment by tomasz-grzesik** [comment]
No, tak, ale dwa pola wprowadzają zamieszanie i sprzedawca może nie wiedzieć, czy spełnił wymaganie, czy nie.
Stąd pytanie, czy w przypadku, gdy oferta zawiera jeden produkt nie będzie to wystarczające?

**Comment by skiewlicz** [comment]
Jak długo dodane załączniki są przechowywane przez allegro. Czy jeden wygenerowany attachmentId można używać wielokrotnie? Czy przesłane załączniki można gdzieś zobaczyć bądź je pobrać?

**Comment by MaciejFrackowiak** [comment]
Tak, jeden uploadowany załącznik może być dołączony do dowolnej liczby ofert/produktów. Wykonując [PUT /sale/offer-attachments/{attachmentId}](https://developer.allegro.pl/documentation/#operation/uploadOfferAttachmentUsingPUT) w odpowiedzi zwracamy ścieżkę, pod którą dostępny jest załącznik.

**Comment by skiewlicz** [comment]
okej, a jak długo te dane będą przechowywane? Chodzi mi o to, że jeśli dodam załącznik to przez jak długi okres czasu będę mógł wystawiać oferty używając tego załącznika.

**Comment by MaciejFrackowiak** [comment]
@skiewlicz Tak długo jak będzie wykorzystywany, tak długo będzie dostępny.

**Comment by grzman** [comment]
Otrzymuję taki komunikat: 
otrzymałem taki komunikat:
```
Obiekt załącznika utworzony, ID załącznika: 4d1e1618-1f28-4a6e-9c61-679209bec80c
d:\Python\python-3.10.0-embed-amd64\cwiczenia\.venv\Lib\site-packages\urllib3\connectionpool.py:1099: InsecureRequestWarning: Unverified HTTPS request is being made to host 'api.allegro.pl'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(
Błąd SSL, ponawiam próbę (1/3): HTTPSConnectionPool(host='api.allegro.pl', port=443): Max retries exceeded with url: /sale/offer-attachments/4d1e1618-1f28-4a6e-9c61-679209bec80c/content (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:2427)')))
d:\Python\python-3.10.0-embed-amd64\cwiczenia\.venv\Lib\site-packages\urllib3\connectionpool.py:1099: InsecureRequestWarning: Unverified HTTPS request is being made to host 'api.allegro.pl'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(
```
Czy coś jest z zabezpieczeniami SSL po stronie API?

**Comment by MaciejFrackowiak** [comment]
@grzman Spojrzałem w logi i wywołujesz zasób, który nie istnieje - "PUT /sale/offer-attachments/4d1e1618-1f28-4a6e-9c61-679209bec80c/content" - usuń "/content".

**Comment by grzman** [comment]
moment sprawdzę, właćnie pobrałem wszystki ID ofert, czy mogę do każdej tej oferty dodać ten sam załącznik?

**Comment by MaciejFrackowiak** [comment]
@grzman Tak, jeden załącznik może być dodany do dowolnej liczby ofert.

**Comment by grzman** [comment]
otrzymałem teraz to:
`Obiekt załącznika utworzony, ID załącznika: 7c60137d-35b2-49e7-98a5-5a31a13e6a7f
d:\Python\python-3.10.0-embed-amd64\cwiczenia\.venv\Lib\site-packages\urllib3\connectionpool.py:1099: InsecureRequestWarning: Unverified HTTPS request is being made to host 'api.allegro.pl'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(
Błąd SSL, ponawiam próbę (1/3): HTTPSConnectionPool(host='api.allegro.pl', port=443): Max retries exceeded with url: /sale/offer-attachments/7c60137d-35b2-49e7-98a5-5a31a13e6a7f (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:2427)')))
d:\Python\python-3.10.0-embed-amd64\cwiczenia\.venv\Lib\site-packages\urllib3\connectionpool.py:1099: InsecureRequestWarning: Unverified HTTPS request is being made to host 'api.allegro.pl'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(
Błąd SSL, ponawiam próbę (2/3): HTTPSConnectionPool(host='api.allegro.pl', port=443): Max retries exceeded with url: /sale/offer-attachments/7c60137d-35b2-49e7-98a5-5a31a13e6a7f (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:2427)')))
d:\Python\python-3.10.0-embed-amd64\cwiczenia\.venv\Lib\site-packages\urllib3\connectionpool.py:1099: InsecureRequestWarning: Unverified HTTPS request is being made to host 'api.allegro.pl'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(
Błąd SSL, ponawiam próbę (3/3): HTTPSConnectionPool(host='api.allegro.pl', port=443): Max retries exceeded with url: /sale/offer-attachments/7c60137d-35b2-49e7-98a5-5a31a13e6a7f (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:2427)')))
Nie udało się przesłać pliku po maksymalnej liczbie prób.
PS D:\Python\python-3.10.0-embed-amd64\cwiczenia> `

**Comment by MaciejFrackowiak** [comment]
@grzman Obecnie nie przekazujesz poprawnej wartości nagłówka Accept - "*/*". Powinno być "application/vnd.allegro.public.v1+json".

**Comment by MaciejFrackowiak** [comment]
@grzman Tu już nie pomogę, ponieważ nie weryfikujemy kodu. Być może któryś z forumowiczów będzie mógł Tobie pomóc.

**Comment by grzman** [comment]
ok.  dziękuję za pomoc, poradziłem sobie już dalej z tym.

**Comment by grzman** [comment]
Chciałbym wrócić do tematu, wczoraj przez kilka godzin ładowałem załczniki o bezpieczeństwie do ofert prawie 40k ofert, na niektórych ofertach zauważyłem komunikat o błędzie, chciałbym sprawdzić które z ofert nie mają załącznika, można to jakoś sprawdzić? , czy jest na to jakiś filtr w liście ofert(nie znalazłem), czy z poziomu API można?

**Comment by MaciejFrackowiak** [comment]
@grzman Z poziomu WWW takiego filtra obecnie nie ma. W API widać to pobierając szczegóły danej oferty - [GET /sale/product-offers/{offerId}](https://developer.allegro.pl/documentation/#operation/getProductOffer).

**Comment by grzman** [comment]
oferty pobieram z tego zasobu:
url_offers = "https://api.allegro.pl/sale/offers"
czy w tym jest informacja o załcznikach? 
bo troch e się różni ten link od tego który podałeś.

**Comment by Maczuga** [comment]
@grzman Niestety listing ofert nie zawiera informacji o tym. Musisz pobierać każdą ofertę indywidualnie.

**Comment by MaciejFrackowiak** [comment]
@grzman Dokładnie tak jak napisał @Maczuga - na liście ofert tej informacji nie znajdziesz.

**Comment by MaciejFrackowiak** [comment]
@grzman Takie pierwsze opcje już się [pojawiły](https://help.allegro.com/sell/pl/a/rozporzadzenie-gpsr-dwie-nowe-opcje-w-zakladce-moj-asortyment-9gZe6zzolHz).

**Comment by Maczuga** [comment]
A ja dopytam - czy dodanie opcji "TEXT" jest w planach do implementacji przed 13 grudnia?

**Comment by MaciejFrackowiak** [comment]
@Maczuga Tak, działamy by stało się to możliwie jak najszybciej.

**Comment by grzman** [comment]
a jeszcze jedna sprawa, czy mogę dla jednego konta uruchomić jednocześnie np. 4 skrypty które będą ładować pliki pdf, mam asortyment w 4 kategoriach w każdej kategorii po około 50 k +/- ofert i do każdj oferty w danej kategorii muszę dodać inny plik z informacją o bezpieczeństwie. jedną kategorię łąduję około 7 h. czy się coś wykrzacczy?

**Comment by MaciejFrackowiak** [comment]
@grzman Jednocześnie możesz mieć otwartych 20 sesji, więc nie powinno nic się wydarzyć.

**Comment by grzman** [comment]
witam ponownie, czy dla tego zasobu jest potrzebny osobny token dostęu?
[GET /sale/product-offers/{offerId}](https://developer.allegro.pl/documentation/#operation/getProductOffer)

**Comment by MaciejFrackowiak** [comment]
@grzman Nie, jeden token "[bearer-token-for-user](https://developer.allegro.pl/documentation/#section/Authentication/bearer-token-for-user)" z nadanymi odpowiednimi scope'ami pozwala obsłużyć wszystkie zasoby naszego API.

**Comment by tomasz-grzesik** [comment]
Rozumiem, że na razie na Allegro nie widać tych załączników i weryfikować można tylko po odczytanej strukturze oferty?
A to coś dotyczy innych załączników? Użytkownik niestety ma problem, bo nie widzi załadowanych załączników, a starsze pole wprowadza zamieszanie:

![obraz](https://github.com/user-attachments/assets/6eee23ae-de91-4537-9514-f292ac781202)

**Comment by MaciejFrackowiak** [comment]
@tomasz-grzesik Zmiany na froncie są developowane poza API, dlatego na chwilę obecną dostępne są tylko w API, w strukturze oferty. Niebawem pojawią się także na stronie WWW.

**Comment by tomasz-grzesik** [comment]
Dzięki za potwierdzenie. Bo gorzej, gdyby Użytkownicy pouzupełniali, a Ty okaże się, że nie tędy droga :)

**Comment by tomasz-grzesik** [comment]
@MaciejFrackowiak pytanie jeszcze dlaczego na tej stronie załączniki radzi się ładować przez dotychczasową funkcjonalność załączników ofertowych?
https://help.allegro.com/sell/pl/a/jakie-obowiazki-naklada-na-ciebie-rozporzadzenie-gpsr-x5xK1MOxac1

![obraz](https://github.com/user-attachments/assets/6f069c41-662e-4502-84d2-48c53f8d8a35)

**Comment by MaciejFrackowiak** [comment]
@tomasz-grzesik W API również jest jeden mechanizm uploadu plików, tylko plik z typem "SAFETY_INFORMATION_MANUAL" możesz później podczepić pod produkt w productSet.

**Comment by tomasz-grzesik** [comment]
No tak. Jeden sposób uploudu, ale dwa różne sposoby dodawania załączników - jeden do oferty, drugi do produktu.
Użytkownik jednak się pyta, czy ładując sposobem takim jak w poradniku będzie spełniał wymagania.
Sugeruję korzystanie z nowego sposobu, ale tych załączników nie widzą w ofercie :)

**Comment by MaciejFrackowiak** [comment]
@tomasz-grzesik Strona WWW jest nadal w przygotowaniu. Na chwilę obecną pełne info jest wyłącznie w API.

**Comment by grzman** [comment]
Jeśli ktoś chętny mogę udostępnić skrypt który działa i ładuje załączniki. Napisany w Pythonie, 
wymagania to tylko plik csv z ID oferty i ścieżka do pliku z załcznikiem.
ja załadowałem 420000 ofert w kilka dni. musiałem wklejać Nowy token , ale sobie poradziłem. 
Proszę pisać, jełśi ktoś chętny.

**Comment by tomasz-grzesik** [comment]
> @tomasz-grzesik Strona WWW jest nadal w przygotowaniu. Na chwilę obecną pełne info jest wyłącznie w API.

Ok. Przekazałem Użytkownikowi. Dziękuję.
